Add crate docs for solana-program

brson opened the pull request:
This adds fairly comprehensive guidance to the solana-program crate docs.

Topics covered:

- The difference between solana-program and solana-sdk
- Where to find examples in the spl
- How solana programs differ from most rust programs
- How to conventionally define a solana program
- On-chain vs off-chain targets
- Core data type descriptions
- Serialization strategies
- CPI
- Native programs, distinction between different kinds, where to find them
- Sysvars
- Limitations to the Solana version of Rust std

I also added some docs to the log and sysvar modules.

After this I intend to work top-down making sure all the links in the crate docs lead to useful information, modules all have short descriptions, etc.

brson commented:
There are two sections here that are elaborations of information that already lives on docs.solana.com

- "Native programs" is like https://docs.solana.com/developing/runtime-facilities/programs
- "Limitations of the Solana implementation of the Rust standard library" is like https://docs.solana.com/developing/on-chain-programs/developing-rust#restrictions

To limit the size of the crate docs I could update those sections of docs.solana.com and have the solana-program crate direct readers to them.

brson commented:
It's not clear what the CI error is.

brson commented:
I think some of my docs for the log module regarding where to see logs are incorrect. Need to investigate more.

CriesofCarrots commented:
> It's not clear what the CI error is.

Looks like the CI agent oomed. I restarted it on a different agent.

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//!   by the Solana runtime.
+//! - They run in a constrained VM environment, and while they do have access to
+//!   the [Rust standard library][std], many features of the standard library,
+//!   particularly related to I/O, will fail at runtime, will silently do
```
How about instead of "I/O" how about "OS Services"

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//! This `cfg` pattern is suitable as well for user code that needs to work both
+//! on-chain and off-chain.
+//!
+//! `solana-program` and `solana-sdk` were previously a single crate. Because of
```
Would be helpful to point out what does and does not work in more detail...or if we can, wes should clean up this discrepancies

brson replied:
Can I do this one in a follow up pr?

jackcmay replied:
Sounds good

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//!
+//! - [`Pubkey`] - An [ed25519] public key. All Solana accounts are identified
+//!   by a `Pubkey`, though some may not have corresponding private keys. As
+//!   Solana programs do not handle secret keys, the full [`Keypair`] is not
```
Not sure about "do not handle", the bigger reason is that programs don't and can't safely create private keys while running and a `Keypair` is never offered to them.  So they don't have any reason to use these.

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//!
+//! # Core data types
+//!
+//! - [`Pubkey`] - An [ed25519] public key. All Solana accounts are identified
```
We often refer in the docs and elsewhere to account "addresses" and the `Pubkey` is how addresses are expressed.  Public key language is more commonly used when paired with a private key.  Often account addresses don't have private keys or the private keys aren't relevant to the use of the address.

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//!   passing it a set of accounts and program-specific data.
+//! - [`ProgramError`] and [`ProgramResult`] - The error type that all programs
+//!   must return, reported to the runtime as a `u64`.
+//!
```
Could also reference types from "native_token.rs"

brson replied:
I'll add the `Sol` type here with some nuance to indicate that the conversions from lamports in native_token are perhaps more important - I haven't seen much code that uses the `Sol` type.

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//!   by a `Pubkey`, though some may not have corresponding private keys. As
+//!   Solana programs do not handle secret keys, the full [`Keypair`] is not
+//!   defined in `solana-program` but in `solana-sdk`.
+//! - [`Hash`] - A [SHA-256] hash, used to uniquely identify blocks.
```
Thge hash type is generic and not only used for blockhashes

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//!   utilities in its [`borsh` module][borshmod] that are not available in the
+//!   `borsh` library.
+//!
+//!   The `Instruction` type has built-in support for borsh.
```
What is this referring to?

brson replied:
The `new_with_borsh` constructor. I'll clarify.

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//! [`Keypair`]: https://docs.rs/solana-sdk/latest/solana_sdk/signer/keypair/struct.Keypair.html
+//! [SHA-256]: https://en.wikipedia.org/wiki/SHA-2
+//!
+//! # Serialization
```
It should probably be pointed out that both borsh and bincode are expensive (with the latter being very) and that developers are free to use whatever serialisation or data format they wish for their own account and instruction data.  But also, they for data they did not source (sysvars for example) they need to deserialize with he correct format

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//! [invoked]: #cross-program-instruction-execution
+//! [transactions]: https://docs.rs/solana-sdk/latest/solana_sdk/transaction/index.html
+//!
+//! __Special built-ins__ include a few programs that are defined like built-ins,
```
What do you consider in this category?

brson replied:
The compute budget program and the zk token proof program. They are the only built-ins I found that don't fall into the precompile or program loader category and also don't work with on-chain programs.

Reading through all the native programs and rationalizing for myself the differences between them was challenging. I'm happy for any suggestions about how to document them and distinguish their differences. I'm happy to rename this category that I made up, or document it in a different way.

brson replied:
From my reading, these native programs could be invoked from other programs but their instructions are defined elsewhere:

- solana_vote_program
- solana_config_program
- address_lookup_table_program

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//! particular program loader then that loader will be used to run the program.
+//! Program loaders are typically invoked by the CLI `solana deploy` command,
+//! to deploy programs, and by the runtime to execute programs. Invoking a
+//! program loader from another program will return an error.
```
you can invoke the loader from another program to issue loader instructions like `Deploy` or `Authorize`, etc...

brson replied:
I did try to invoke the bpf_loader program with the Write instruction and it failed. I'll give it another try and figure out what I'm missing.

jackcmay replied:
Let me know what you encounter

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//! to deploy programs, and by the runtime to execute programs. Invoking a
+//! program loader from another program will return an error.
+//!
+//! This crate defines the program IDs for most native programs: even though
```
This could be clarified.  Some program's are not invokable (like precompiles) but a program needs access to the program id, for example, to verify that a precompile instruction was included in the transaction that this program is a part of

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//! instructions they process, and constructors for building the instructions.
+//! Some built-ins' instructions and constructors, though technically callable
+//! from Solana programs, are not defined somewhere accessible to Solana
+//! programs. This is a historical artifact due to lack of use cases for
```
Which are you referring to, maybe we should move then into this crate

brson replied:
From my reading, these native programs could be invoked from other programs but their instructions are defined elsewhere:

- solana_vote_program
- solana_config_program
- address_lookup_table_program

jackcmay replied:
Programs don't need to be visible to other programs as programs won't/shouldn't be calling them.

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//!
+//! [lut]: https://docs.solana.com/proposals/transactions-v2
+//!
+//! ## Special built-ins
```
The compute budget program can probably just be lumped into the precompiled list.  And actually we could probably refer to these as all builtins and then specify which are program invokable and which are only supported as top-level instructions

jackcmay commented on sdk/program/src/lib.rs:
````diff
@@ -1,3 +1,650 @@
+//! }
+//! ```
+//!
+//! As an optimization, to avoid including sysvar accounts in instructions and deserializing them
````
This is the preferred way and should be highlighted first.  The account method of getting sysvars should be clearly described as a way to deal older programs that still pass sysvars as accounts.

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//!
+//! [sysvardoc]: https://docs.solana.com/developing/runtime-facilities/sysvars
+//!
+//! # Limitations to the Solana version of the Rust standard library
```
Would be best just to refer to docs.solana.com.  Any suggestions you have to improve the docs would be appreciated, the doc src files are here: https://github.com/solana-labs/solana/tree/master/docs/src

brson commented:
Thanks for all the great feedback @jackcmay. I'm reworking.

brson commented:
@jackcmay I've pushed a commit that addresses your feedback. Notes:

- I would like to document behavioral differences in solana-program between on- and off-chain programs in a different PR as it is a lot of effort.
- I added the `Sol` type to the core data types section
- I completely reworked the "native programs" section.
- I updated the text to indicate that program loaders are callable on-chain, though I have failed to call them in my own experiments.
- I deleted the section on Solana modifications to std and will follow up sometime with enhancements to the relevant page on docs.solana.com.

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,560 @@
+//!   [_lamports_], the smallest fractional unit of SOL, in the [`native_token`]
+//!   module.
+//!
+//! [acc]: https://docs.solana.com/developing/programming-model/accounts
```
What's `acc`?

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,560 @@
+//! [ed25519]: https://ed25519.cr.yp.to/
+//! [`Keypair`]: https://docs.rs/solana-sdk/latest/solana_sdk/signer/keypair/struct.Keypair.html
+//! [SHA-256]: https://en.wikipedia.org/wiki/SHA-2
+//! [`Sol`]: native_token::Sol
```
Why are the types duplicated here?

jackcmay commented on sdk/program/src/lib.rs:
```diff
@@ -1,3 +1,650 @@
+//!
+//! The native programs are:
+//!
+//! ## Built-ins
```
Don't think the address lookup table either

brson replied:
Removed.

jackcmay commented:
Looking good, tiny nits

Love to hear more about the difficulty you had calling the loader

brson commented:
I removed the address lookup table program from the docs.

I won't be able to experiment with invoking the bpf loader again until next week, but otherwise I think I've addressed everything.

jackcmay commented:
@Mergifyio rebase

jackcmay commented:
Thanks for the great contribution!
